src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java

```diff
@@ -0,0 +1,472 @@
+         * (3) Paired ends are grouped by a similar key as above but using both reads.
+         * (4) Any unpaired end is emitted, unmodified.
+         * (5) The remained paired ends are scored and all but the highest scoring are marked as
+         *     duplicates. Both reads in the pair are emitted.
+         */
+        private PCollection<Read> transformReads(final PCollectionView<SAMFileHeader> headerPcolView, final PCollection<Read> pairs) {
+            final PCollection<KV<String, Iterable<Read>>> pairsGroupedByKey = groupPairsByKey(pairs, headerPcolView);
+            final PTransform<PCollection<? extends KV<String, Iterable<Read>>>, PCollection<KV<String, PairedEnds>>> markGroupedDuplicatePairs = markGroupedDuplicatePairs(headerPcolView);
+            final PTransform<PCollection<? extends KV<String, Iterable<PairedEnds>>>, PCollection<Read>> markPairedEnds = markPairedEnds();
+
+            return pairsGroupedByKey
+                    .apply(markGroupedDuplicatePairs)
+                    .setCoder(KvCoder.of(StringUtf8Coder.of(), PairedEndsCoder.of()))
+                    .apply(GroupByKey.<String, PairedEnds>create())
+                    .apply(markPairedEnds);
+        }
+
+        /**
+         * Groups pairs by keys - keys are tuples of (library, contig, position, orientation).
+         */
+        private PCollection<KV<String, Iterable<Read>>> groupPairsByKey(final PCollection<Read> pairs, final PCollectionView<SAMFileHeader> headerPcolView) {
+            final PTransform<PCollection<? extends Read>, PCollection<KV<String, Read>>> makeKeysForPairs =
+                    ParDo.named("make keys for pairs")
+                            .withSideInputs(Collections.singletonList(headerPcolView))
+                            .of(new DoFn<Read, KV<String, Read>>() {
+                                private static final long serialVersionUID = 1l;
+
+                                @Override
+                                public void processElement(final ProcessContext context) throws Exception {
+                                    final Read record = context.element();
+                                    if (GenomicsReadUtils.isPaired(record)) {
+                                        final SAMFileHeader h = context.sideInput(headerPcolView);
+                                        final String key = MarkDuplicatesReadsKey.keyForPair(h, record);
+                                        final KV<String, Read> kv = KV.of(key, record);
+                                        context.output(kv);
+                                    }
+                                }
+                            });
+            return pairs.apply(makeKeysForPairs).apply(GroupByKey.<String, Read>create());
+        }
+
+        private PTransform<PCollection<? extends KV<String, Iterable<Read>>>, PCollection<KV<String, PairedEnds>>> markGroupedDuplicatePairs(final PCollectionView<SAMFileHeader> headerPcolView) {
```

Comment:
Here and below: For these PTransforms, you can define the transform and use them as follow:
pCollectionOfStuff.apply(new Transform1(headerSingleton)).apply(new Transform2(headerSingleton)...

You don't need to define methods to return the transforms.